nav2-usability/scenario-code/sign-in-routing/router.dart

```diff
@@ -0,0 +1,309 @@
+}
+
+class AppState extends ChangeNotifier {
+  final Authentication auth;
+
+  AppState(this.auth);
+
+  Future<bool> signIn(String username, String password) async {
+    var success = await auth.signIn(username, password);
+    notifyListeners();
+    return success;
+  }
+
+  bool _viewingBooks = false;
+
+  bool get viewingBooks => _viewingBooks;
+
+  set viewingBooks(bool value) {
+    _viewingBooks = value;
+    notifyListeners();
+  }
+}
+
+class _BooksAppState extends State<BooksApp> {
+  final BookRouterDelegate _routerDelegate = BookRouterDelegate();
+  final BookRouteInformationParser _routeInformationParser =
+  BookRouteInformationParser();
+
+  @override
+  void dispose() {
+    _routerDelegate.dispose();
+    super.dispose();
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return MaterialApp.router(
+      title: 'Books App',
+      routerDelegate: _routerDelegate,
+      routeInformationParser: _routeInformationParser,
+    );
+  }
+}
+
+class BookRouteInformationParser extends RouteInformationParser<AppRoutePath> {
+  @override
+  Future<AppRoutePath> parseRouteInformation(
+      RouteInformation routeInformation) async {
+    final uri = Uri.parse(routeInformation.location!);
+
+    // Handle '/'
+    if (uri.pathSegments.isEmpty) {
+      return HomeRoutePath();
+    }
+
+    if (uri.pathSegments.length == 1) {
+      // Handle '/signin'
+      if (uri.pathSegments[0] == 'signin') return SignInRoutePath();
+      if (uri.pathSegments[0] == 'books') return BooksRoutePath();
+    }
+
+    // Handle unknown routes
+    return HomeRoutePath();
+  }
+
+  @override
+  RouteInformation restoreRouteInformation(AppRoutePath path) {
+    late final String location;
+    if (path is HomeRoutePath) {
+      location = '/';
+    } else if (path is BooksRoutePath) {
+      location = '/books';
+    } else if (path is SignInRoutePath) {
+      location = '/signin';
+    }
+    return RouteInformation(location: location);
+  }
+}
+
+class BookRouterDelegate extends RouterDelegate<AppRoutePath>
+    with ChangeNotifier, PopNavigatorRouterDelegateMixin<AppRoutePath> {
+  @override
+  final GlobalKey<NavigatorState> navigatorKey;
+  final AppState _appState = AppState(MockAuthentication());
+
+  BookRouterDelegate() : navigatorKey = GlobalKey<NavigatorState>() {
+    _appState.addListener(() => notifyListeners());
+  }
+
+  @override
+  void dispose() {
+    _appState.dispose();
+    super.dispose();
+  }
+
+  @override
+  AppRoutePath get currentConfiguration {
+    if (!_appState.auth.isSignedIn()) {
+      return SignInRoutePath();
+    } else if (_appState.viewingBooks) {
+      return BooksRoutePath();
+    } else {
+      return HomeRoutePath();
+    }
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    final isSignedIn = _appState.auth.isSignedIn();
+    final viewingBooksScreen = _appState.viewingBooks;
+    return Navigator(
+      key: navigatorKey,
+      pages: [
+        if (isSignedIn) ...[
```

Comment:
sign checking isSignedIn is usually an async operation in web application, we should do it during the route parsing

Reply:
Let me see if this works - making `isSignedIn` async could cause some problems with `currentConfiguration`